docs/developer-feature-tutorial.md

```diff
@@ -99,74 +165,113 @@ Now, run zulip terminal and open web app from a different account. Start composi
+    """
+    Handle typing notifications (in private messages)
+    """
+    assert event['type'] == "typing"
+    if hasattr(self.controller, 'view'):
         # If the user is in pm narrow with the person typing
-        if len(self.model.narrow) == 1 and\
-                self.model.narrow[0][0] == 'pm_with' and\
-                event['sender']['email'] in self.model.narrow[0][1].split(','):
+        narrow = self.narrow
+        if (len(narrow) == 1 and narrow[0][0] == 'pm_with' and event['sender']['email'] in narrow[0][1].split(',')):
             if event['op'] == 'start':
-                user = self.model.user_dict[event['sender']['email']]
-                self._w.footer.set_text([
+                user = self.user_dict[event['sender']['email']]
+                self.controller.view.set_footer_text([
                     ' ',
                     ('code', user['full_name']),
                     ' is typing...'
                 ])
-                self.controller.update_screen()
             elif event['op'] == 'stop':
-                self._w.footer.set_text(self.get_random_help())
-                self.controller.update_screen()
+                self.controller.view.set_footer_text()
+            else:
+                raise RuntimeError("Unknown typing event operation")
 ```
-If the conditions are satisfied, we display `x is typing` if the `op` is `start` and display help message if the `op` is `stop`.
+
+The above code uses the`set_footer_text()` method implemented in the `View` class in `ui.py`, which should default to setting a random Help message when not being used.
+
+If the conditions are satisfied, we display `<sender> is typing...` if the `op` is `start` and display a help message if the `op` is `stop`.
 
 There are two parts to updating a widget in urwid:
-* Changing the widget
-* Updating the screen
 
-This line of code,
-```python
-self._w.footer.set_text([
-                    ' ',
-                    ('code', user['full_name']),
-                    ' is typing...'
-                ])
-```
-changes the footer text, and this
+* Changing the widget, which is done by View
+* Updating the screen, which is done by Controller
+
+These are done by the highlighted lines accordingly in `set_footer_text`
+
 ```python
-self.controller.update_screen()
+@asynch
+def set_footer_text(self, text_list: Optional[List[Any]]=None, duration: Optional[float]=None) -> None:
+    if text_list is None:
+        text = self.get_random_help()
+    else:
+        text = text_list
+>>> self._w.footer.set_text(text)
+>>> self.controller.update_screen()
+    if duration is not None:
+        assert duration > 0
+        time.sleep(duration)
+        self.set_footer_text()
 ```
-updates the screen to display the changes. This fully implements the typing feature.
 
-### Writing tests
+## Writing tests
+
+We use `pytest` to write all our tests. We have to write new tests, modify current ones or delete old ones whenever needed.
```

Comment:
I'd include some extra notes here; "have to" doesn't really motivate people :)

I think we may have notes in the README now, but if we update code, we can say we expect the tests to be adapted to pass in the same commit, and ideally add new tests for new features (or test cases for extended behavior).

There are examples of both here that you can point to :)

Another motivation for getting linting & tests working well could be to introduce how later they will be run on branches and pull requests on pushing to github, which will give them an "official" pass/fail result!